readme.md
## Express middleware for basic authentication

### Prerequisites
* [express](https://expressjs.com) based webserver
* mysql ([knex](https://knexjs.org) compatible) database (see @pubcore/knex-auth)

### Features
* supports two type of users: SYSTEM and HUMAN
* will set an user object to express req object (if authentication succeeded)
* serves allways "401 Unauthorized" and cancel URI, if no credentials
* serves allways "401 Unautherized" and cancel URI, if user not found
* calls express "next", if username and password is ok
* redirect to "deactivated" page, if wrong password used too much within a time window (105ms)
* redirect to "deactivated" page, if last login of user is long time ago
* updates last login stamp, one time within defined time frame
* redirects to change password page (including a back-uri), on first request of new user
* redirects to change password page (including a back-uri), if password is expired
* support secondary password for SYSTEM users
* does set a flag to user object (oldPwUsed), if secondary password exists, but old password has been used
* optinal support of login by JsonWebToken cookie (Jwt), enabled if option "jwtKeyFile" is available

[activity diagram](doc/authentication-flow.pdf)

### Configuration options (set on server startup)
		options = {
			publicCancelLoginUri:'/login/canceled',
			publicDeactivatedUri:'/login/deactivated',
			changePasswordUri:'/login/pwchange',
			maxTimeWithoutActivity: 1000 * 60 * 60 * 24 * 180,//[msec]
			maxLoginAttempts:10,
			maxLoginAttemptsTimeWindow:1000 * 3600 * 24,//[msec]
			minTimeBetweenUpdates:1000 * 3600,//[msec],
			jwtKeyFile:'/run/secret/jwt-key.txt' //optional
		},
		table = 'user',

### Example
		const
			createLoginMiddleware = require('@pubcore/express-basic-auth').default,
			options = {
				changePasswordUri:'/login/pwchange',
				publicDeactivatedUri:'/login/deactivated',
				publicCancelLoginUri:'/login/canceled',
				maxTimeWithoutActivity: 1000 * 60 * 60 * 24 * 180,//[msec]
				maxLoginAttempts:10,
				maxLoginAttemptsTimeWindow:1000 * 3600 * 24,//[msec]
				minTimeBetweenUpdates:1000 * 3600,//[msec]
			},
			table = 'user',
			knex = new Knex({
				client: 'mysql', connection: {/* see knex*/}
			}),
			db = {knex, table}

		const login = createLoginMiddleware({db, options})
		var router = express.Router()
		router.all('/', login)


CHANGELOG.md
## 1.1.0 2019-05-22
* introduce support login by Json Web Token cookie (Jwt)
* remove transpile stuff

test/index.spec.js
'use strict'

const chai = require('chai'),
	{expect} = chai,
	login = require('../js/index').default,
	chaiHttp = require('chai-http'),
	express = require('express'),
	{dbTypes} = require('@pubcore/knex-auth'),
	createTestDb = require('@pubcore/knex-create-test-db').default,
	defaultMap = require('./userDefaultMap').default,
	{resolve} = require('path'),
	cookie = require('cookie'),
	JWT = require('jsonwebtoken')

chai.use(chaiHttp)
const app = express(),
	app2 = express(),
	options = {
		publicDeactivatedUri:'/login/deactivated',
		changePasswordUri:'/login/pwchange',
		publicCancelLoginUri:'/login/canceled',
		maxTimeWithoutActivity: 1000 * 60 * 60 * 24 * 180,//[msec]
		maxLoginAttempts:2,
		maxLoginAttemptsTimeWindow:50,//[msec]
		minTimeBetweenUpdates:500,//[msec]
	},
	table = 'user',
	knex = createTestDb({table, rows:defaultMap([{}, {
		username:'adam', password:null, last_login:null,
		password_new:'tempPw', created_time:new Date(),
		password_expiry_date:new Date()
	}, {
		username:'bob', last_login:new Date('2000-01-01T00:00:00')
	}, {
		username:'tom', password_secondary:'new'
	}, {
		username:'ben', password_expiry_date:new Date()
	}]), beforeEach, after, dbTypes:{
		...dbTypes, first_name:'string', last_name:'string', email:'string'
	}}),
	db = {knex, table},
	error = err => {throw err},
	Jwt = JWT.sign({ username: 'eve' }, 'somestring')

app.use(login({db, options}))
app.use('/', (req, res) => res.send(req.user))

//JWT support for app2
app2.use((req, res, next) => {
	req.cookies = cookie.parse(req.headers.cookie || '')
	next()
})
app2.use(login({db, options:{...options, jwtKeyFile:resolve(__dirname, 'jwtKey.txt')}}))
app2.use('/', (req, res) => res.send(req.user))

var lastMediaType = 0
const aMediaType = () => {
	return (['text/html', 'application/json', 'application/xml'])[++lastMediaType%3]
}
const repead = (n, p) => {
	return Promise.all((new Array(n)).map(() => p))
}
const expect401 = res => {
	expect(res).to.have.status(401)
	expect(res.text).to.contain(options.publicCancelLoginUri)
}
const expect200 = res => expect(res).to.have.status(200)
const wait = ms => new Promise(res => setTimeout(()=>res(), ms))
const wrongPasswordRequest = () =>
	chai.request(app).get('/').set('Accept', aMediaType()).redirects(0).auth('eve', 'xyz')
const correctPasswordRequest = (username='eve') =>
	chai.request(app).get('/').redirects(0).auth(username, 'test')

describe('http authentication service', () => {
	it('exports API', () => expect(login).not.to.be.undefined)
	it('serves allways "401 Unauthorized" and cancel URI, if no credentials',
		() => repead(3, chai.request(app).get('/').then(expect401))
	)
	it('serves allways "401 Unautherized" and cancel URI, if user not found',
		() => repead(3, chai.request(app).get('/').auth('xyz', 'test').then(expect401))
	)
	it('serves "200 ok" if username and password is ok',
		() => correctPasswordRequest().then(expect200, error)
	)
	it('redirect to "deactivated" page, if wrong password used too much within a time window',
		() => wrongPasswordRequest().then(expect401)
			//after a correct request counter is reset
			.then(() => correctPasswordRequest().then(expect200))
			.then(() => wrongPasswordRequest().then(expect401))
			//wait to get outsite of time window
			.then(() => wait(options.maxLoginAttemptsTimeWindow))
			.then(() => wrongPasswordRequest().then(expect401))
			.then(() => wrongPasswordRequest().then(
				res => expect(res).redirectTo(options.publicDeactivatedUri)
			))
			//now, even a request with correct password must lead to deactivated
			.then(() => correctPasswordRequest().then(
				res => expect(res).redirectTo(options.publicDeactivatedUri)
			))
	)
	it('redirect to "deactivated" page, if last login of user is long time ago',
		() => chai.request(app).get('/').redirects(0).auth('bob', 'test')
			.then(res => expect(res).redirectTo(options.publicDeactivatedUri))
	)
	it('updates last login stamp, one time within defined time frame', () => {
		var firstStamp
		return correctPasswordRequest().then(({body}) => {
			firstStamp = body.last_login
		}, error)
			.then(() => correctPasswordRequest().then(({body}) => {
				expect(firstStamp).to.equal(body.last_login)
			}, error))
			.then(() => wait(options.minTimeBetweenUpdates))
			.then(() => correctPasswordRequest())
			.then(() => correctPasswordRequest().then(({body}) => {
				expect(firstStamp).to.not.equal(body.last_login)
			}, error))
	})
	it('redirects to change password page (including a back-uri), on first request of new user', () => {
		var uri = '/xyz/?foo=bar'
		return chai.request(app).get(uri).redirects(0).auth('adam', 'tempPw').then(
			res => expect(res)
				.redirectTo(options.changePasswordUri)
				.and.to.have.cookie('back-uri', encodeURIComponent(uri))
		)
	})
	it('redirects to change password page (including a back-uri), if password is expired', () => {
		var uri = '/foo/?bar=xyz'
		return chai.request(app).get(uri).redirects(0).auth('ben', 'test').then(
			res => expect(res)
				.redirectTo(options.changePasswordUri)
				.and.to.have.cookie('back-uri', encodeURIComponent(uri))
		)
	})
	it('does set a flag (oldPwUsed), if secondary password exists, but old password has been used', () =>
		correctPasswordRequest('tom').then(({body}) => expect(body.oldPwUsed).to.be.true)
	)
	it('does not add any password fields', () =>
		correctPasswordRequest('tom').then(({body}) =>
			expect(JSON.stringify(body)).to.not.contain('password')
		)
	)
	it('serves basic user data', () =>
		correctPasswordRequest('tom').then(({body}) =>
			expect(JSON.stringify(body)).to.contain('first_name')
				.and.contain('last_name').and.contain('email')
		)
	)
	it('supports (optional) login by Json Web Token (JWT); depends on jwtKeyFile option exists (staticly cached)', () =>
		chai.request(app2).get('/').set('Cookie', cookie.serialize('Jwt', Jwt)).redirects(0).then(
			expect200, error
		)
	)
	it('rejects with 401 Unautherized, if JWT is invalid', () =>
		chai.request(app2).get('/').set('Cookie', cookie.serialize('Jwt', Jwt+'garbleIt')).redirects(0).then(
			expect401, error
		)
	)
})


test/userDefaultMap.js
export default (users=[{}]) => users.map(usr => ({...{
	username:'eve',
	type:'HUMAN',
	deactivate:'no',
	login_failed_count:0,
	password:'$2y$04$RG9G38B8UD2zoHnP9MD2eunCc6hJxvmVly5r/1CRg9el3kfptw8Ra',
	last_login: new Date(),
	first_name: 'Ursa',
	last_name: 'Minor',
	email:'um@xy.com'
}, ...usr}))


js/lib/authenticate.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _authentication = require('@pubcore/authentication');

var _authentication2 = _interopRequireDefault(_authentication);

var _authUser = require('@pubcore/auth-user');

var _authenticateGofer = require('../gofer/authenticateGofer');

var _authenticateGofer2 = _interopRequireDefault(_authenticateGofer);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

exports.default = function (_ref) {
	var username = _ref.username,
	    password = _ref.password,
	    getUser = _ref.getUser,
	    getOptions = _ref.getOptions,
	    rest = _objectWithoutProperties(_ref, ['username', 'password', 'getUser', 'getOptions']);

	return (0, _authentication2.default)({
		username: username,
		password: password,
		gofer: (0, _authenticateGofer2.default)(rest),
		carrier: { getUser: getUser, getOptions: getOptions },
		lib: { comparePassword: _authUser.comparePassword }
	});
};

js/gofer/authenticateGofer.js
'use strict'
const {deactivateUser, addLoginFailed, resetLoginFailedCount,
		updateLastLogin} = require('@pubcore/knex-auth'),
	http401 = require('../lib/http401').default,
	cookie = require('cookie')

exports.default = ({db, res, req, options}) => {
	var {publicDeactivatedUri, changePasswordUri, publicCancelLoginUri} = options

	return {
		noCredentials: () => http401({publicCancelLoginUri, res}),
		notFound: () => http401({publicCancelLoginUri, res}),
		isDeactivated: () =>
			req.path !== publicDeactivatedUri && res.redirect(publicDeactivatedUri),
		toDeactivate: ({username}) => deactivateUser(db, {username}).then(
			() => res.redirect(publicDeactivatedUri)
		),
		invalidWebToken: () => http401({publicCancelLoginUri, res}),
		invalidPassword: ({username}) => addLoginFailed(db, {username}).then(
			() => http401({publicCancelLoginUri, res})
		),
		authenticated: (user, isTimeToUpdate) => {
			var {login_failed_count, username} = user
			return Promise.resolve(
				login_failed_count > 0 && resetLoginFailedCount(db, {username})
			).then(
				() => isTimeToUpdate && updateLastLogin(db, {username})
			).then(() => user)
		},
		oldPwUsed: user => (user.oldPwUsed = true) && user,
		passwordExpired: () => {
			res.setHeader(
				'Set-Cookie',
				cookie.serialize('back-uri', String(req.originalUrl), {
					httpOnly: true, path:'/', secure:true
				}))
			req.path !== changePasswordUri && res.redirect(changePasswordUri)
		}
	}
}


js/index.js
'use strict'
const basicAuth = require('basic-auth'),
	knexAuth = require('@pubcore/knex-auth'),
	getUser = knexAuth.default,
	{comparePassword} = knexAuth,
	authenticate = require('@pubcore/authentication').default,
	gofer = require('./gofer/authenticateGofer').default,
	{readFile} = require('fs'),
	authenticateOptions = { //all time values in [ms]
		maxTimeWithoutActivity: 1000 * 60 * 60 * 24 * 180,
		maxLoginAttempts:5,
		maxLoginAttemptsTimeWindow:1000 * 60 * 60 * 24,
	},
	httpOptions = {
		changePasswordUri:'/login/pwChange',
		publicDeactivatedUri:'/login/deactivated',
		publicCancelLoginUri:'/login/canceled',
	},
	loadSecret = file => new Promise((resolve, reject) => readFile(
		file,
		{encoding:'utf8'},
		(err, data) => err ? reject(err) : resolve(data.trim())
	))

var jwtKey = ''

exports.default = ({db, options}) => (...args) => {
	var [req, res, next] = args,
		{name, pass} = basicAuth(req) || {},
		{cookies} = req,
		{Jwt} = cookies || {}

	return authenticate({
		jwt:Jwt,
		username:name,
		password:pass,
		gofer:gofer({
			db, req, res, options:{...httpOptions, ...options}
		}),
		carrier: {
			getOptions: () => Promise.resolve(
				options.jwtKeyFile && !jwtKey && loadSecret(options.jwtKeyFile)
					.then(data => jwtKey = data, err => Promise.reject(err))
			).then(() => ({...authenticateOptions, ...options, jwtKey})),
			getUser:({username}) => getUser(
				{...db, cols:['first_name', 'last_name', 'email']}, {username}
			)
		},
		lib:{comparePassword}
	}).then(user => {
		if(user){
			var {username, email, first_name, last_name, last_login, oldPwUsed} = user
			req.user = {
				username, email, first_name, last_name, last_login, oldPwUsed
			}
		}
		next()
	}).catch(next)
}
